File: LeetCode_Algorithm/39(combianation_sum)/39.js
var solve = function(candidates_sliced, path, target, ans) {
  if (target < 0) {
    return ;
  }
  if (target == 0) {
    ans.push(path); return ;
  }

  for (let i in candidates_sliced) {
    let path_temp = path.slice(); path_temp.push(candidates_sliced[i])
    solve(candidates_sliced.slice(i, ),path_temp, target-candidates_sliced[i], ans)
  }
}

var combinationSum = function(candidates, target) {
  // sol1 204ms 7%
  candidates.sort((a,b) => {
    return a-b
  })
  let ans = []
  for (let i in candidates) {
    let temp = candidates.slice(i, )
    solve(temp, [candidates[i]], target-candidates[i], ans)
  }
  return ans
};

console.log(combinationSum([2,3,5], 8))

File: LeetCode_Algorithm/39(combianation_sum)/39.py
from typing import List


class Solution:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        # sol1 288ms 8%
        if target < 0: return []
        def solve(candidates, path, target, ans):
            if target < 0:
                return 
            if target == 0:
                ans.append(path); return
            for i in range(len(candidates)):
                solve(candidates[i:], path+[candidates[i]], target-candidates[i], ans)
            pass
        ans = []
        for i in range(len(candidates)):
            solve(candidates[i:], [candidates[i]], target-candidates[i], ans)
        return ans


        pass

print(Solution().combinationSum([2,3,5], 8))
